### PR TITLE
Implement stdmodule clang detection flag

### PR DESCRIPTION
**Pull request by Arthapz**
draft because the flag got re-reverted a second time :D but it should be backported to llvm18 this week, see https://github.com/llvm/llvm-project/pull/82160

### CODE REVIEW

File: xmake/rules/c++/modules/modules_support/clang/compiler_support.lua [hunk 1]
```diff
@@ -72,6 +72,26 @@ function _get_cpplibrary_name(target)
     end
 end
 
+function _get_std_module_manifest_path(target)
+    local print_module_manifest_flag = get_printlibrarymodulemanifestpathflag(target)
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang/compiler_support.lua:76]: oh no, maybe we should add `_` for get_printlibrarymodulemanifestpathflag

File: xmake/rules/c++/modules/modules_support/clang/compiler_support.lua [hunk 2]
```diff
@@ -72,6 +72,26 @@ function _get_cpplibrary_name(target)
...
+    if print_module_manifest_flag then
+        local outdata = try { function() return os.iorun(self:program() .. " " .. print_module_manifest_flag) end }
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang/compiler_support.lua:78]: please use os.iorunv

File: xmake/rules/c++/modules/modules_support/clang/compiler_support.lua [hunk 3]
```diff
@@ -73,9 +73,10 @@ function _get_cpplibrary_name(target)
 end
 
 function _get_std_module_manifest_path(target)
-    local print_module_manifest_flag = get_printlibrarymodulemanifestpathflag(target)
+    local print_module_manifest_flag = get_print_library_module_manifest_path_flag(target)
     if print_module_manifest_flag then
-        local outdata = try { function() return os.iorun(self:program() .. " " .. print_module_manifest_flag) end }
+        local compinst = target:compiler("cxx")
+        local outdata, _ = try { function() return os.iorunv(compinst:program(), {print_module_manifest_flag}) end }
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang/compiler_support.lua:79]: {envs = compinst:runenvs()}

**Comment by waruqi** [comment]
Is it ok now?

File: xmake/rules/c++/modules/modules_support/clang/compiler_support.lua [hunk 4]
```diff
@@ -239,15 +260,7 @@ function get_stdmodules(target)
...
-                -- @see https://github.com/llvm/llvm-project/pull/76451 (has been revert, so we need to wait)
-                local clang_path = path.directory(get_clang_path(target))
-                local clang_lib_path = path.join(clang_path, "..", "lib")
-                local modules_json_path = path.join(clang_lib_path, "libc++.modules.json")
-                if not os.isfile(modules_json_path) then
-                    modules_json_path = find_file("*/libc++.modules.json", clang_lib_path)
-                end
+                local modules_json_path = _get_std_module_manifest_path(target)
                 if modules_json_path then
                     local modules_json = json.decode(io.readfile(modules_json_path))
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang/compiler_support.lua:265]:
```
if modules_json and modules_json.modules and #modules_json.modules > 0 then
```

**Comment by Arthapz** [comment]
> s it ok now?
seems like it has been relanded https://github.com/ChuanqiXu9/llvm-project/commit/27665923ce0b9974601d7e5880506e3560bade19

**Comment by Arthapz** [comment]
The PR got merged, but i don't get it to work on my setup (see https://github.com/llvm/llvm-project/pull/84881#issuecomment-2002529955) so i'll pass this to ready when i confirmed it's working

**Comment by Arthapz** [comment]
i tried it and it work, fallback work too

**Comment by waruqi** [comment]
Thanks!

**Pull request merged**